cilium/CFP-25694-move-cilium-cli-to-cilium-repo.md

```diff
@@ -0,0 +1,70 @@
+# CFP-25694: Move cilium/cilium-cli code into cilium/cilium repository
+
+**Code Name: Mega Ketchup**
+
+**SIG: [sig-cilium-cli]**
+
+**Begin Design Discussion:** 2023-05-25
+
+**Cilium Release:** 1.15
+
+**Authors:** Michi Mutsuzaki `<michi@isovalent.com>`
+
+## Summary
+
+Move [cilium/cilium-cli] code into [cilium/cilium] repository.
+
+## Motivation
+
+The main motivation for moving [cilium/cilium-cli] code into [cilium/cilium]
+repository is to minimize the feature development overhead for:
+
+- adding new tests to `cilium connectivity test`.
+- adding new tasks to `cilium sysdump`.
+
+## Goals
+
+* Move [cilium/cilium-cli] code into [cilium/cilium] main branch under `cilium-cli/` directory.
+* Establish a backporting process for changes inside `cilium-cli/` directory.
+* Don't disrupt the cilium-cli development and release processes.
+
+## Non-Goals
+
+* Rename the name of the `cilium-cli` binary from `cilium` to something else.
+* Release `cilium-cli` as a part of Cilium releases.
+
+## Proposal
+
+### Overview
+
+### Releasing cilium-cli
+
+We do not need to change the release process / cadence to accomplish the goals
+defined in this CFP. Continue releasing `cilium-cli` from the main branch at its
+own cadence using its own versioning that is independent of [cilium/cilium]
+versioning as we have been doing in [cilium/cilium-cli] repository.
+
+We need to come up with a git tag convention for cilium-cli releases to distinguish
+tags for Cilium releases from tags for cilium-cli releases. My proposal to use
+`cilium-cli/vX.Y.Z` for cilium-cli release tags.
```

Comment:
Whatever scheme we come up with, this needs to be compatible with Go modules, otherwise it will break users who pull in cilium/cilium or cilium/cilium-cli as a go module.

If we want to merge merge cilium-cli into the `github.com/cilium/cilium` Go module, then pulling in cilium-cli as a dependency will be tricky, because then the version of the Go module would inherently be tied to Cilium (the CNI) releases, because Go mod requires semantic versioning, and `cli-v0.13.1` is not a valid semantic version.

If, on the other hand, we want to keep cilium-cli as it's own Go module, and it's import path would be e.g. `github.com/cilium/cilium/cli` then we'll need something like `cli/vX.Y.Z` according to https://go.dev/blog/publishing-go-modules - but that means we'd still a dependency from the CLI module to the CNI module. That might require us to use Go workspaces.

I think this point (separate Go module vs. shared Go module) should be a key question that needs to be answered as part of the CFP.

Reply:
... i have another radical idea. let me try and see if it works 🧠

Reply:
Another option might be to use an approach similar to what upstream k8s is doing for some of its modules, e.g. `k8s.io/apimachinery` where code changes are made in the main k8s repo (i.e. https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apimachinery in case of `k8s.io/apimachinery`) and these are then synced to their own repo (i.e. https://github.com/kubernetes/apimachinery in case of `k8s.io/apimachinery`).

I think using a similar approach for cilium-cli would allow to use the in-tree version for cilium/cilium CI while still allowing independent releases from cilium/cilium-cli without having to keep e.g. multiple `go.mod` files in cilium/cilium or using Go workspaces.

Reply:
yeah i think this is the cleanest way to avoid impacting users who import `github.com/cilium/cilium-cli` as a module.

i was thinking about a similar, but much lazier approach:

- https://github.com/cilium/design-cfps/blob/bf48480707bb3b31ed182b21754b1d50748cbc35/cilium/CFP-25694-move-cilium-cli-to-cilium-repo.md?plain=1#L57-L70
- https://github.com/cilium/design-cfps/blob/bf48480707bb3b31ed182b21754b1d50748cbc35/cilium/CFP-25694-move-cilium-cli-to-cilium-repo.md?plain=1#L84-L93

i'm fine with either of these approaches. i'll document both 📝

Reply:
... on the second thought, i like your approach better. let me just document your approach.

Reply:
... i'm having third thought 😹 if we take this approach, where does `go.mod` file for `cilium-cli` come from? i see that these staging directories contain `go.mod` files with some `replace` directives. for example: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/go.mod

Reply:
> not keeping a separate go.mod for the in-tree cilium-cli at all and just use the main go.mod. 

yeah i don't think keeping cilium-cli as a separate module is worth all the effort. 

however, with this approach, you also need to modify all the import paths in .go files back from `github.com/cilium/cilium/cli` to `github.com/cilium/cilium-cli`.

... i'm starting to like my original proposal again 😹